src/components/fields/AnyOfField.js

```diff
@@ -30,7 +30,45 @@ class AnyOfField extends Component {
 
   getMatchingOption(formData, options) {
     for (let i = 0; i < options.length; i++) {
-      if (isValid(options[i], formData)) {
+      const option = options[i];
+
+      // If the schema describes an object then we need to add slightly more
+      // strict matching to the schema, becuase unless the schema usies the
+      // "requires" keyword, an object will match the schema as long as it
+      // doesn't have matching keys witha conflicting type. To do this we use an
+      // "anyOf" with an array of requires. This augmentation expresses that the
+      // schema should match if any of the keys in the schema are present on the
+      // object and pass validation.
+      if (option.properties) {
+        // Create an "anyOf" schema that requires at least on of the keys in the
+        // "properties" object
+        const requiresAnyOf = {
+          anyOf: Object.keys(option.properties).map(key => ({
+            required: [key],
+          })),
+        };
+
+        let augmentedSchema;
+
+        // If the "anyOf" keyword already exists, wrap it in an "allOf"
+        if (option.anyOf) {
+          const { anyOf, ...rest } = option;
+
+          if (!rest.allOf) {
+            rest.allOf = [];
+          }
+
+          rest.allOf.push(...[requiresAnyOf, { anyOf }]);
```

Comment:
I think we should be careful not to mutate `rest.allOf` (because `rest.allOf` is a shallow copy of `option.allOf`).

I think `arr.push(...[a, b])` is the same as `arr.push(a, b)`. I don't think we need to add the existing `anyOf` to `allOf` -- each schema keyword is implicitly `allOf`-ed, right?

Reply:
Good catch on the shallow copy, I'll fix that and add a test that checks for mutation.
You're right, the spread operator in `.push` is unnecessary - I'll remove it.
Regarding the `allOf/anyOf` merging, I read back through the code, and you're correct, its not necessary in this case. I think this is a result of some refactoring I was doing to simplify the code paths, I'll refactor again and clean it up.